**Polars**

1. Считайте датасет из файла train.csv

In [60]:
import polars as pl
df = pl.read_csv('train.csv')

2. Выведите основную информацию о датасете: информацию о типах данных, число пропусков, средние значения и т.д.

In [3]:
df.describe()

shape: (9, 13)
┌────────────┬─────────────┬──────────┬──────────┬───┬───────────┬───────────┬───────┬──────────┐
│ statistic  ┆ PassengerId ┆ Survived ┆ Pclass   ┆ … ┆ Ticket    ┆ Fare      ┆ Cabin ┆ Embarked │
│ ---        ┆ ---         ┆ ---      ┆ ---      ┆   ┆ ---       ┆ ---       ┆ ---   ┆ ---      │
│ str        ┆ f64         ┆ f64      ┆ f64      ┆   ┆ str       ┆ f64       ┆ str   ┆ str      │
╞════════════╪═════════════╪══════════╪══════════╪═══╪═══════════╪═══════════╪═══════╪══════════╡
│ count      ┆ 891.0       ┆ 891.0    ┆ 891.0    ┆ … ┆ 891       ┆ 891.0     ┆ 204   ┆ 889      │
│ null_count ┆ 0.0         ┆ 0.0      ┆ 0.0      ┆ … ┆ 0         ┆ 0.0       ┆ 687   ┆ 2        │
│ mean       ┆ 446.0       ┆ 0.383838 ┆ 2.308642 ┆ … ┆ null      ┆ 32.204208 ┆ null  ┆ null     │
│ std        ┆ 257.353842  ┆ 0.486592 ┆ 0.836071 ┆ … ┆ null      ┆ 49.693429 ┆ null  ┆ null     │
│ min        ┆ 1.0         ┆ 0.0      ┆ 1.0      ┆ … ┆ 110152    ┆ 0.0       ┆ A10   ┆ C        │
│ 25%        ┆ 224.0       ┆ 0.0      ┆ 2.0      ┆ … ┆ null      ┆ 7.925     ┆ null  ┆ null     │
│ 50%        ┆ 446.0       ┆ 0.0      ┆ 3.0      ┆ … ┆ null      ┆ 14.4542   ┆ null  ┆ null     │
│ 75%        ┆ 669.0       ┆ 1.0      ┆ 3.0      ┆ … ┆ null      ┆ 31.0      ┆ null  ┆ null     │
│ max        ┆ 891.0       ┆ 1.0      ┆ 3.0      ┆ … ┆ WE/P 5735 ┆ 512.3292  ┆ T     ┆ S        │
└────────────┴─────────────┴──────────┴──────────┴───┴───────────┴───────────┴───────┴──────────┘

In [4]:
df.dtypes

[Int64,
 Int64,
 Int64,
 String,
 String,
 Float64,
 Int64,
 Int64,
 String,
 Float64,
 String,
 String]

In [5]:
df.null_count()

shape: (1, 12)
┌─────────────┬──────────┬────────┬──────┬───┬────────┬──────┬───────┬──────────┐
│ PassengerId ┆ Survived ┆ Pclass ┆ Name ┆ … ┆ Ticket ┆ Fare ┆ Cabin ┆ Embarked │
│ ---         ┆ ---      ┆ ---    ┆ ---  ┆   ┆ ---    ┆ ---  ┆ ---   ┆ ---      │
│ u32         ┆ u32      ┆ u32    ┆ u32  ┆   ┆ u32    ┆ u32  ┆ u32   ┆ u32      │
╞═════════════╪══════════╪════════╪══════╪═══╪════════╪══════╪═══════╪══════════╡
│ 0           ┆ 0        ┆ 0      ┆ 0    ┆ … ┆ 0      ┆ 0    ┆ 687   ┆ 2        │
└─────────────┴──────────┴────────┴──────┴───┴────────┴──────┴───────┴──────────┘

In [6]:
df.mean()

shape: (1, 12)
┌─────────────┬──────────┬──────────┬──────┬───┬────────┬───────────┬───────┬──────────┐
│ PassengerId ┆ Survived ┆ Pclass   ┆ Name ┆ … ┆ Ticket ┆ Fare      ┆ Cabin ┆ Embarked │
│ ---         ┆ ---      ┆ ---      ┆ ---  ┆   ┆ ---    ┆ ---       ┆ ---   ┆ ---      │
│ f64         ┆ f64      ┆ f64      ┆ str  ┆   ┆ str    ┆ f64       ┆ str   ┆ str      │
╞═════════════╪══════════╪══════════╪══════╪═══╪════════╪═══════════╪═══════╪══════════╡
│ 446.0       ┆ 0.383838 ┆ 2.308642 ┆ null ┆ … ┆ null   ┆ 32.204208 ┆ null  ┆ null     │
└─────────────┴──────────┴──────────┴──────┴───┴────────┴───────────┴───────┴──────────┘

3. Посчитайте количество пассажиров каждого класса (Pclass)

In [66]:
df.get_column("Pclass").value_counts()

shape: (3, 2)
┌────────┬───────┐
│ Pclass ┆ count │
│ ---    ┆ ---   │
│ i64    ┆ u32   │
╞════════╪═══════╡
│ 3      ┆ 491   │
│ 2      ┆ 184   │
│ 1      ┆ 216   │
└────────┴───────┘

4. Выведите количество выживших мужчин и женщин на корабле

In [65]:
df.group_by("Sex").agg(pl.col("Survived").sum().alias("Survived_Count"))

shape: (2, 2)
┌────────┬────────────────┐
│ Sex    ┆ Survived_Count │
│ ---    ┆ ---            │
│ str    ┆ i64            │
╞════════╪════════════════╡
│ female ┆ 233            │
│ male   ┆ 109            │
└────────┴────────────────┘

5. Выведите часть таблицы с пассажирами, возраст которых больше 44 лет

In [9]:
df.filter(pl.col("Age") > 44)

shape: (115, 12)
┌─────────────┬──────────┬────────┬──────────────────┬───┬────────────┬─────────┬───────┬──────────┐
│ PassengerId ┆ Survived ┆ Pclass ┆ Name             ┆ … ┆ Ticket     ┆ Fare    ┆ Cabin ┆ Embarked │
│ ---         ┆ ---      ┆ ---    ┆ ---              ┆   ┆ ---        ┆ ---     ┆ ---   ┆ ---      │
│ i64         ┆ i64      ┆ i64    ┆ str              ┆   ┆ str        ┆ f64     ┆ str   ┆ str      │
╞═════════════╪══════════╪════════╪══════════════════╪═══╪════════════╪═════════╪═══════╪══════════╡
│ 7           ┆ 0        ┆ 1      ┆ McCarthy, Mr.    ┆ … ┆ 17463      ┆ 51.8625 ┆ E46   ┆ S        │
│             ┆          ┆        ┆ Timothy J        ┆   ┆            ┆         ┆       ┆          │
│ 12          ┆ 1        ┆ 1      ┆ Bonnell, Miss.   ┆ … ┆ 113783     ┆ 26.55   ┆ C103  ┆ S        │
│             ┆          ┆        ┆ Elizabeth        ┆   ┆            ┆         ┆       ┆          │
│ 16          ┆ 1        ┆ 2      ┆ Hewlett, Mrs.    ┆ … ┆ 248706     ┆ 16.0    ┆ null  ┆ S        │
│             ┆          ┆        ┆ (Mary D          ┆   ┆            ┆         ┆       ┆          │
│             ┆          ┆        ┆ Kingcome…        ┆   ┆            ┆         ┆       ┆          │
│ 34          ┆ 0        ┆ 2      ┆ Wheadon, Mr.     ┆ … ┆ C.A. 24579 ┆ 10.5    ┆ null  ┆ S        │
│             ┆          ┆        ┆ Edward H         ┆   ┆            ┆         ┆       ┆          │
│ 53          ┆ 1        ┆ 1      ┆ Harper, Mrs.     ┆ … ┆ PC 17572   ┆ 76.7292 ┆ D33   ┆ C        │
│             ┆          ┆        ┆ Henry Sleeper    ┆   ┆            ┆         ┆       ┆          │
│             ┆          ┆        ┆ (My…             ┆   ┆            ┆         ┆       ┆          │
│ …           ┆ …        ┆ …      ┆ …                ┆ … ┆ …          ┆ …       ┆ …     ┆ …        │
│ 858         ┆ 1        ┆ 1      ┆ Daly, Mr. Peter  ┆ … ┆ 113055     ┆ 26.55   ┆ E17   ┆ S        │
│             ┆          ┆        ┆ Denis            ┆   ┆            ┆         ┆       ┆          │
│ 863         ┆ 1        ┆ 1      ┆ Swift, Mrs.      ┆ … ┆ 17466      ┆ 25.9292 ┆ D17   ┆ S        │
│             ┆          ┆        ┆ Frederick Joel   ┆   ┆            ┆         ┆       ┆          │
│             ┆          ┆        ┆ (Ma…             ┆   ┆            ┆         ┆       ┆          │
│ 872         ┆ 1        ┆ 1      ┆ Beckwith, Mrs.   ┆ … ┆ 11751      ┆ 52.5542 ┆ D35   ┆ S        │
│             ┆          ┆        ┆ Richard Leonard… ┆   ┆            ┆         ┆       ┆          │
│ 874         ┆ 0        ┆ 3      ┆ Vander Cruyssen, ┆ … ┆ 345765     ┆ 9.0     ┆ null  ┆ S        │
│             ┆          ┆        ┆ Mr. Victor       ┆   ┆            ┆         ┆       ┆          │
│ 880         ┆ 1        ┆ 1      ┆ Potter, Mrs.     ┆ … ┆ 11767      ┆ 83.1583 ┆ C50   ┆ C        │
│             ┆          ┆        ┆ Thomas Jr (Lily  ┆   ┆            ┆         ┆       ┆          │
│             ┆          ┆        ┆ A…               ┆   ┆            ┆         ┆       ┆          │
└─────────────┴──────────┴────────┴──────────────────┴───┴────────────┴─────────┴───────┴──────────┘

**Ускорение** **работы с pandas**

1. Считайте датасет из файла train.csv

In [10]:
import pandas as pd
df_pandas = pd.read_csv('train.csv')

2. Посчитайте средний возраст пассажиров и его стандартное отклонение с помощью bottleneck

In [11]:
import bottleneck as bn
bn.nanmean(df_pandas['Age'])

29.69911764705882

In [12]:
bn.nanstd(df_pandas['Age'])

14.516321150817317

3. Для каждого пассажира умножьте значение столбца Fare на 1.3 и сохраните результаты как новый столбец Fare_new

In [44]:
df = df.with_columns((pl.col("Fare") * 1.3).alias("Fare_new"))
print(df.select(["Fare", "Fare_new"]).head())

shape: (5, 2)
┌─────────┬──────────┐
│ Fare    ┆ Fare_new │
│ ---     ┆ ---      │
│ f64     ┆ f64      │
╞═════════╪══════════╡
│ 7.25    ┆ 9.425    │
│ 71.2833 ┆ 92.66829 │
│ 7.925   ┆ 10.3025  │
│ 53.1    ┆ 69.03    │
│ 8.05    ┆ 10.465   │
└─────────┴──────────┘


**Оптимизация типов pandas**

1. Считайте датасет из файла Housing.csv

In [27]:
import numpy as np
df_housing = pd.read_csv('Housing.csv')

2. Для каждого столбца определите оптимальный с точки зрения потребления памяти тип данных

In [22]:
df_housing.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 545 entries, 0 to 544
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   price             545 non-null    int64 
 1   area              545 non-null    int64 
 2   bedrooms          545 non-null    int64 
 3   bathrooms         545 non-null    int64 
 4   stories           545 non-null    int64 
 5   mainroad          545 non-null    object
 6   guestroom         545 non-null    object
 7   basement          545 non-null    object
 8   hotwaterheating   545 non-null    object
 9   airconditioning   545 non-null    object
 10  parking           545 non-null    int64 
 11  prefarea          545 non-null    object
 12  furnishingstatus  545 non-null    object
dtypes: int64(6), object(7)
memory usage: 55.5+ KB


In [29]:
df_housing.describe()

price        area    bedrooms   bathrooms     stories  \
count    545.000000  545.000000  545.000000  545.000000  545.000000   
mean    -323.317431   -0.460550    2.965138    1.286239    1.805505   
std    19211.594898   77.272477    0.738064    0.502470    0.867492   
min   -32552.000000 -128.000000    1.000000    1.000000    1.000000   
25%   -16624.000000  -72.000000    2.000000    1.000000    1.000000   
50%    -1304.000000    0.000000    3.000000    1.000000    2.000000   
75%    16404.000000   66.000000    3.000000    2.000000    2.000000   
max    32644.000000  127.000000    6.000000    4.000000    4.000000   

          parking  
count  545.000000  
mean     0.693359  
std      0.861328  
min      0.000000  
25%      0.000000  
50%      0.000000  
75%      1.000000  
max      3.000000

3. Поменяйте типы данных столбцов датафрейма на выбранные вами в прошлом пункте и сравните потребление памяти до и после оптимизации

In [28]:
df_housing=df_housing.fillna(0)
df_housing['price']=df_housing['price'].astype(np.int16)
df_housing['area']=df_housing['area'].astype(np.int8)
df_housing['bedrooms']=df_housing['bedrooms'].astype(np.int8)
df_housing['bathrooms']=df_housing['bathrooms'].astype(np.int8)
df_housing['stories']=df_housing['stories'].astype(np.int8)
df_housing['parking']=df_housing['parking'].astype(np.float16)
df_housing.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 545 entries, 0 to 544
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   price             545 non-null    int16  
 1   area              545 non-null    int8   
 2   bedrooms          545 non-null    int8   
 3   bathrooms         545 non-null    int8   
 4   stories           545 non-null    int8   
 5   mainroad          545 non-null    object 
 6   guestroom         545 non-null    object 
 7   basement          545 non-null    object 
 8   hotwaterheating   545 non-null    object 
 9   airconditioning   545 non-null    object 
 10  parking           545 non-null    float16
 11  prefarea          545 non-null    object 
 12  furnishingstatus  545 non-null    object 
dtypes: float16(1), int16(1), int8(4), object(7)
memory usage: 34.2+ KB


Оптимизация числовых типов данных позволила уменьшить размер массива с 55.5 Кбайт до 34.2 Кбайт